### DISCUSSION TITLE
how to specify the correct type for the Controller

### DISCUSSION THREAD

**Discussion by ya2s** [react-hook-form/react-hook-form#7674] [Q&A] [answered]
I get a type error when I try with the sample code below

```
import { Controller, Control } from "react-hook-form";

type FieldValue1 = {
  value1: string;
};
type FieldValue2 = {
  value2: string;
};

type FormData = FieldValue1 | FieldValue2;

type Props<T extends FormData> = {
  control: Control<T>;
};

export const Component = <T extends FormData>({ control }: Props<T>) => {
  return (
    <Controller
      control={control}
      render={({ field: { value } }) => <>{value}</>}
      name="test"
      defaultValue=""
    />
  );
};

```

> Type '"test"' is not assignable to type 'Path<T>'

Error message

<img width="852" alt="CleanShot 2022-01-27 at 16 16 49@2x" src="https://user-images.githubusercontent.com/42740055/151309871-7b4a5596-c810-4886-a47b-443df6ab46cc.png">

How can I specify the correct type?

**Comment by Moshyfawn** [comment, marked as answer]
You've specified your form values types as 
```js
type FieldValue1 = {
  value1: string;
};
type FieldValue2 = {
  value2: string;
};
```
which means, that your from field name can be either `value1` or `value2`, as there's no `test` field specified in your type.
Here's what Typescript autocomplete suggests as a valid field name inferring from your types"
![image](https://user-images.githubusercontent.com/16290753/151322053-ac92ea03-a94e-4b86-8efd-cf7af9a310c9.png)

You might wanna do this instead:
```js
type FromData = {
  test: FieldValue1 | FieldValue2
}
```
So, that way you can specify your field name as `name='test.value1'`

**Reply by ya2s** [reply]
@Moshyfawn Thank you for the answer!
It's my mistake! It should be as follows

```
type FieldValue1 = {
  test: string;
  value1: string;
};
type FieldValue2 = {
  test: string;
  value2: string;
};
```

The error still occurs

**Reply by felixschorer** [reply]
@yamatsum Typescript is unable to resolve `Path<T>` due to the generic and the complexity of the type. That's a limitation of type level programming in TS. You'll have to use a type assertion.

```ts
const control = props.control as Control<FormData>
```

**Reply by ya2s** [reply]
@felixschorer 
Using type assertions doesn't seem to solve the problem
<img width="1362" alt="CleanShot 2022-01-29 at 22 53 54@2x" src="https://user-images.githubusercontent.com/42740055/151663696-0b211160-fc47-473d-ab49-ed67fe0d0e86.png">

**Reply by felixschorer** [reply]
My bad, you need use `as never as Control<FormData>`.

**Reply by Moshyfawn** [reply]
Not `as unknown`?

**Reply by felixschorer** [reply]
For a type assertion to succeed, the types need to be related in some way. `unknown` is a super type of every type, `never` is a subtype of every type. It doesn't really matter, both should work.

**Reply by Moshyfawn** [reply]
Fair enough, thx for the explanation!
